File: 2018/day_01/day_01.rb
#! ruby

require 'set'

def part_2(drifts)
    seen = Set.new
    freq = 0
    drifts.cycle do |delta|
        freq += delta
        break freq unless seen.add?(freq)
    end
end

drifts = File.new("day_01.input").each_line.map(&:to_i)

puts "Part 1: #{drifts.sum}"
puts "Part 2: #{part_2(drifts)}"

File: 2018/day_01/nim_01/day_01.nim
import strutils
import sequtils
import math
import intsets

proc part_2(drifts: seq): int =
  var seen = initIntSet()
  var acc = 0
  while true:
    for i in drifts:
      acc = acc + i
      if seen.contains(acc):
        return acc
      else:
        seen.incl(acc)

let drifts = readFile("../day_01.input").strip.splitlines.map(parseInt)

echo "Part 1: ", drifts.sum()
echo "Part 2: ", part_2(drifts)
